semver/range.ts

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+import { comparatorIntersects, SemVerComparator } from "./comparator.ts";
+import { INVALID, SemVer } from "./semver.ts";
+import { gte, lte, sort } from "./operators/mod.ts";
+
+type SemVerRangeAnd = SemVerComparator[];
+type SemVerRangeOr = SemVerRangeAnd[];
+
+/**
+ * A type representing a semantic version range. The ranges consist of
+ * a nested array, which represents a set of OR comparisons while the
+ * innner array represents AND comparisons.
+ */
+export interface SemVerRange {
+  // The outter array is OR while each inner array is AND
+  ranges: SemVerRangeOr;
+}
+
+/**
+ * The minimum valid SemVer for a given range or INVALID
+ * @param range The range to calculate the min for
+ * @returns A valid SemVer or INVALID
+ */
+export function rangeMin(range: SemVerRange): SemVer { // For and's, you take the biggest min
+  // For or's, you take the smallest min
+  //[ [1 and 2] or [2 and 3] ] = [ 2 or 3 ] = 2
+  return sort(
+    range.ranges.map((r) =>
+      sort(r.filter((c) => rangeTest(c.min, range)).map((c) => c.min)).pop()!
+    ).filter((v) => v),
+  ).shift() ?? INVALID;
+}
+
+/**
+ * The maximum valid SemVer for a given range or INVALID
```

Comment:
Why INVALID instead of `undefined`?

Reply:
good question, I'll think about it and either alter it or give you a good answer.